File: ios/LowerBackStretching/Data/Achievements.swift
import Foundation

enum AchievementId: String, CaseIterable {
    case firstSession
    case sevenDayStreak
    case thirtyDayStreak
    case fiftySessions
    case hundredSessions
    case levelFive
    case levelTen
}

struct Achievement: Equatable {
    let id: AchievementId
    let title: String
    let description: String
    /// N events required. 1 = binary unlock.
    let target: Int
}

struct AchievementStatus: Equatable {
    let achievement: Achievement
    let progress: Int
    let unlocked: Bool
}

enum Achievements {
    static let all: [Achievement] = [
        Achievement(id: .firstSession,     title: "First steps",   description: "Complete your first stretching session.", target: 1),
        Achievement(id: .sevenDayStreak,   title: "On a roll",     description: "Stretch 7 days in a row.",                target: 7),
        Achievement(id: .thirtyDayStreak,  title: "Habit forged",  description: "Stretch 30 days in a row.",               target: 30),
        Achievement(id: .fiftySessions,    title: "Half a hundred",description: "Complete 50 sessions.",                   target: 50),
        Achievement(id: .hundredSessions,  title: "Centurion",     description: "Complete 100 sessions.",                  target: 100),
        Achievement(id: .levelFive,        title: "Limber",        description: "Reach level 5.",                          target: 5),
        Achievement(id: .levelTen,         title: "Bendy",         description: "Reach level 10.",                         target: 10),
    ]
}

/// Compute the status of every achievement from already-derived stats
/// (no DB access). Locked rows still carry a `progress` value so the UI
/// can render a "23 / 50" indicator under each badge.
func evaluateAchievements(
    totalSessions: Int,
    longestStreak: Int,
    level: Int
) -> [AchievementStatus] {
    Achievements.all.map { a in
        let raw: Int
        switch a.id {
        case .firstSession, .fiftySessions, .hundredSessions:
            raw = totalSessions
        case .sevenDayStreak, .thirtyDayStreak:
            raw = longestStreak
        case .levelFive, .levelTen:
            raw = level
        }
        return AchievementStatus(
            achievement: a,
            progress: min(raw, a.target),
            unlocked: raw >= a.target
        )
    }
}

File: android/app/src/main/java/com/lowerbackstretching/data/Achievements.kt
package com.lowerbackstretching.data

enum class AchievementId {
    FIRST_SESSION,
    SEVEN_DAY_STREAK,
    THIRTY_DAY_STREAK,
    FIFTY_SESSIONS,
    HUNDRED_SESSIONS,
    LEVEL_FIVE,
    LEVEL_TEN,
}

data class Achievement(
    val id: AchievementId,
    val title: String,
    val description: String,
    /** N events required. 1 = binary unlock. */
    val target: Int,
)

data class AchievementStatus(
    val achievement: Achievement,
    val progress: Int,
    val unlocked: Boolean,
)

object Achievements {
    val all: List<Achievement> = listOf(
        Achievement(AchievementId.FIRST_SESSION, "First steps", "Complete your first stretching session.", 1),
        Achievement(AchievementId.SEVEN_DAY_STREAK, "On a roll", "Stretch 7 days in a row.", 7),
        Achievement(AchievementId.THIRTY_DAY_STREAK, "Habit forged", "Stretch 30 days in a row.", 30),
        Achievement(AchievementId.FIFTY_SESSIONS, "Half a hundred", "Complete 50 sessions.", 50),
        Achievement(AchievementId.HUNDRED_SESSIONS, "Centurion", "Complete 100 sessions.", 100),
        Achievement(AchievementId.LEVEL_FIVE, "Limber", "Reach level 5.", 5),
        Achievement(AchievementId.LEVEL_TEN, "Bendy", "Reach level 10.", 10),
    )
}

/**
 * Compute the status of every achievement from already-derived stats
 * (no DB access). Locked rows still carry a [progress] value so the UI
 * can render a "23 / 50" indicator under each badge.
 */
fun evaluateAchievements(
    totalSessions: Int,
    longestStreak: Int,
    level: Int,
): List<AchievementStatus> = Achievements.all.map { a ->
    val raw = when (a.id) {
        AchievementId.FIRST_SESSION,
        AchievementId.FIFTY_SESSIONS,
        AchievementId.HUNDRED_SESSIONS -> totalSessions
        AchievementId.SEVEN_DAY_STREAK,
        AchievementId.THIRTY_DAY_STREAK -> longestStreak
        AchievementId.LEVEL_FIVE,
        AchievementId.LEVEL_TEN -> level
    }
    AchievementStatus(
        achievement = a,
        progress = raw.coerceAtMost(a.target),
        unlocked = raw >= a.target,
    )
}
